bpo-30386: Add a build infrastructure for Android.

xdegaye opened the pull request:
Add a build infrastructure for Android to cross-compile Python using the Android NDK and allowing to run python interactively on the emulator or to run the Python test suite using the Android SDK.


<!-- issue-number: bpo-30386 -->
https://bugs.python.org/issue30386
<!-- /issue-number -->

xdegaye committed:
bpo-30386: Add a build infrastructure for Android.

AraHaan commented:
Mind also doing this for iOS as well so people can build their own free version of python for the iOS. It is annoying having to pay $$$ for 3.4+ on it but nothing for 2.7...

Also what about pip on the android and iOS? I thought they might not get pip to work on them.

xdegaye committed:
bpo-30386: Support the new NDK Unified Headers.

xdegaye committed:
Add Makefile targets to implement the buildbot steps.

xdegaye committed:
Merge branch 'master' into bpo-30386

xdegaye committed:
Update the native interpreter build on Setup.dist and version changes.

xdegaye commented:
Commit details can be found at issue bpo-30386.

xdegaye committed:
Add build a distribution for an application on a device

xdegaye commented:
>  xdegaye requested a review from tiran as a code owner 5 hours ago

I did not request a review from tiran.

xdegaye committed:
The device/emulator distributions include now only the shared libraries from the external libraries

xdegaye committed:
The WITH_LIBRARIES environment variable selects the external libraries

xdegaye committed:
Add the whatsnew and Misc/NEWS.d entries

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-30386

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-30386

xdegaye committed:
Fix a regression from 3161893d933538f4bd93c674e52cc8043885694e

terminfo/l/linux is now again installed on the emulator.

xdegaye committed:
Patch libffi to set FFI_MMAP_EXEC_WRIT in configure

See https://github.com/libffi/libffi/pull/265.
The external libraries are now untared in the build tree instead
of the source tree previously.

xdegaye committed:
Java 8 is now required to create an avd.

This is at Android build-tools version 25.0.3 and emulator version 26.0.3.

pmp-p commented:
@xdegaye is there some dedicated place to discuss about android ports since you are not on irc ?

xdegaye commented:
@pmp-p you can open a new issue on the [issue tracker](https://devguide.python.org/tracker/) and if you want to involve me in the discussion just add ``xdegaye`` to the nosy list of this issue, or you can send an email to [python-ideas or python-dev](https://devguide.python.org/communication/).
It depends of what you want to discuss.

xdegaye committed:
Build ctypes with the headers of the cross-compiled libffi

xdegaye committed:
Disable _ctypes on x86_64 (all long double tests fail)

xdegaye committed:
Disable uuid, Android has no uuid/uuid.h header

xdegaye committed:
Disable grp, some grp.h functions are __INTRODUCED_IN_FUTURE

xdegaye committed:
Disable _crypt, Android does not have crypt.h

xdegaye committed:
Install also the README(s), test_normalization needs the test/data directory

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-30386

vstinner commented:
> @xdegaye is there some dedicated place to discuss about android ports since you are not on irc ?

Isn't https://mail.python.org/mailman/listinfo/mobile-sig the appropriate mailing list to discuss Android and embedded Python on mobile platforms?

xdegaye committed:
Add the EMULATOR_CMD_LINE_OPTIONS environment variable

Remove the deprecated '-enable-kvm' qemu option.

xdegaye committed:
The default for ANDROID_API is now 24

xdegaye committed:
Disable _ctypes on arm64 platforms

Only one test fails:
  File "/sdcard/org.python/lib/python3.7/ctypes/test/test_win32.py",
        line 137, in test_struct_by_value
    self.assertEqual(ret.left, left.value)
  AssertionError: -200 != 10

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-30386

xdegaye committed:
Do not update Modules/Setup if it already lists the disabled modules

xdegaye committed:
Remove the support of the arm architecture

The supported architectures are now armv7, arm64, x86 and x86_64.

xdegaye committed:
Do not use an SD card partition image (not anymore mounted on 26.1.4)

xdegaye committed:
Fix the build of libffi on armv7

See https://github.com/libffi/libffi/issues/191

xdegaye committed:
Add support for multiple emulators running concurrently

xdegaye committed:
Add the pythoninfo make target

xdegaye committed:
Update the build system to run on buildbots

Change the default SDK and NDK root path from /opt to $HOME/android.

xdegaye committed:
Exit with the return code of the shell run remotely by adb

xdegaye committed:
Run pythoninfo on the emulator and add the NDK version

The logic for starting the emulator has been modified to adjust to the
behavior expected by a buildbot.

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-30386

xdegaye committed:
Call pybuildbot.identify in the buildbottest target

TMPDIR is now the tmp/ subdirectory of $SYS_EXEC_PREFIX

xdegaye committed:
Print the external libraries versions

xdegaye committed:
Do not set now (after bpo-32059) '--sysroot' in CFLAGS

xdegaye committed:
Remove useless CFLAGS options

xdegaye committed:
Update the README

xdegaye committed:
'--sysroot' is needed in CFLAGS by distutils

xdegaye committed:
Do not run nl_langinfo.sh

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-30386

vstinner commented on Android/README.rst:
```diff
@@ -0,0 +1,447 @@
+.. contents::
+
+This document provides a quick overview of the Python build system for Android
+on a **linux platform** [1]_.
```
This documentation is uncomplete. You need "quick start" instructions to explain what is Android, what do I require on my local host, do I need a physical device, etc. This documentation mostly explains the implementation, not really how to use it.

vstinner replied:
I'm trying to run Python on Android in an emulator using these recipes, but it's hard to understand how to use it. First, I downloaded NDK r16b but configure-android *immediately* fails because it *requires* NDK 14... This thing be very explicit in the doc, since the NDK is a big file (800 MB, so slow to download) and takes a lot of disk space (2.6 GB). It's not easy to install it.

xdegaye replied:
I will add a ``quick start`` section shortly.

> First, I downloaded NDK r16b but configure-android immediately fails because it requires NDK 14.

I understand your frustation. The README does say that ``android-ndk-r14 is required`` in a note and that note is referred to in each ``Requirements`` section where the NDK is required. I will add a warning about this in the ``quick start`` section unless you have a better idea.

xdegaye committed:
There is now only one method to build Python

xdegaye committed:
Support openssl-1.1.0g on all architectures

xdegaye committed:
Support ctypes also on 64 bits architectures

xdegaye committed:
Support pip

sys.prefix must not be located on /sdcard: pip adds a sys.prefix/bin/ directory
and fails upon PermissionError raised when using the /sdcard file system.

xdegaye committed:
Remove the Doc/whatsnew/3.7rst and Misc/NEWS.d entries

xdegaye committed:
Set the python version in build.mk instead of the Makefile

xdegaye committed:
Do not use a symlink for inputrc

xdegaye committed:
Add the 'setup' target to build third-party extension modules
